Sources/SwiftDiagnostics/DiagnosticDecorators/DiagnosticDecorator.swift

```diff
@@ -0,0 +1,72 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Protocol that defines a standard interface for decorating diagnostic output in source code.
+///
+/// This protocol is intended to be used by entities such as ``DiagnosticsFormatter`` and ``GroupedDiagnostics``
+/// to apply custom decorations to diagnostic messages, buffer outlines, and code highlights.
+///
+/// ## Conforming to `DiagnosticDecorator`:
+///
+/// To conform to the `DiagnosticDecorator` protocol, you must implement three required methods:
+///
+/// 1. `decorateMessage(_:basedOnSeverity:)`: For decorating diagnostic messages.
+/// 2. `decorateBufferOutline(_:)`: For decorating the outlines of source code buffers.
+/// 3. `decorateHighlight(_:)`: For decorating individual highlights within a source code snippet.
+///
+/// ## Customization:
+///
+/// The protocol is designed to be easily customizable. Developers can create their own entities that conform
+/// to `DiagnosticDecorator` to implement custom decorating logic. This allows for different visual representations,
+/// such as using ANSI colors, underscores, emoji-based or other markers, for diagnostics in source code.
+@_spi(Testing) public protocol DiagnosticDecorator {
```

Comment:
I think we don’t need to make `DiagnosticDecorator` `@_spi(Testing) public` yet, we can just keep it internal (`ANSIDiagnosticDecorator` needs to say `@_spi(Testing)`).

And just to clarify: I think that making `DiagnosticDecorator` might be a very good idea in the future. I would just like to get some experience with the API by implementing a few more diagnostic decorators internally before deciding that we are happy enough with the API to make it public.

Reply:
Done

Reply:
I've modified the implementation to enable error throwing. Let me know what you think.